### ISSUE TITLE
4.8.2 regression: type narrowing bug

### ISSUE THREAD

**Issue by lll000111** [microsoft/TypeScript#50706]
# Bug Report

Instead of words, just the code. The comments say what is wrong and what used to work in 4.7.4.

```javascript
// Arguments are set to "unknown" because they receive data from network requests
// so runtime checks are mandatory to narrow type before use.
function SendBlob(data: unknown, encoding: unknown) {
    if (encoding !== undefined && encoding !== 'utf8') {
        throw new Error('encoding');
    }

    // Mouse hover to see detected type 
    // When using TS v4.8.2 it is {} | undefined
    // When using TS v4.7.4 it is "utf8" | undefined
    encoding;
};
```

### 🔎 Search Terms

"4.8.2" and I found a few but was told to file a new issue because I'm told it wasn't similar to the ones that I found and fought were similar.

I thought e.g. #50671 looks similar, not for the code but possibly for the deeper cause, since I don't think _that_ many things changed in exactly this part of TS since 4.7.4.

### 🕗 Version & Regression Information

4.7.4 works

4.8.2 doesn't work

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=4.8.2#code/GYVwdgxgLglg9mABAZQKZgCYCEA2cBGAFBgIZQkBci4A1mHAO5gA0i6EcGMYA5lbfSYBKRAG8AsAChEMxDGCJC7Ttx6IAhAF5N1TKmDdUGRADITbSCt4btiAOQgowABx2RE6bK9QAFgCdGRDBUBkQAUT8Av0I7ZS5eNwBuKS8AXykU2QB6LMQAWTgQAGdURB84ADdUP0QoOEQS0oxUKFRoI1qATwAHUsyZHMQAdR90aiLVRAAVZEQKgBYAOmdFgCY5KDkisVTEAB9dZoNgjH7EQZGx4smZuaWAdkX5ja3EACJHFzf9w-1DU88MjiqmSklSoKAA

**Comment by andrewbranch** [comment]
Hah, this is a fun one. Switch the order of your checks:

```ts
if (encoding !== 'utf8' && encoding !== undefined) {
```

Narrowing `unknown` has several special rules. A new one in 4.8 is that `unknown` can break down into a union of `{} | null | undefined`. That’s triggered as soon as you filter `undefined` out of it in your first check, and at that point, you’re left with `{} | null`, and so the next narrowing step can only discriminate between those two things. But if you reverse the order, a different set of special rules around strict equality in general takes effect first.

**Comment by lll000111** [comment]
Ah I have another one, single undefined check condition:

https://www.typescriptlang.org/play?ts=4.8.2#code/GYVwdgxgLglg9mABAEzgZTgWwKZQBYxgDmAPACqLYAeU2YyAzouANZhwDuYAfABQBuAQwA2IbAC5EZAJSSKAbwCwAKERrEMYIgEixiALyHm9bMELZk0xEtXq7AJ1wh7SIaOwBuFXYC+K7+qa2m56hvqIACwATFY2dnb49pyIAKL2Sfa8AOQAstiCYIREiHBawprYiCHYWdJetmp+ygFqjlDOrrqeKk1AA

```javascript
function doSomething<T extends unknown>(value: T): T {
    if (value === undefined) {
        return value;
    }

    // ERROR: This condition will always return false because...
    if (value === 42) {
        throw Error('Meaning of life value');
    }

    return value;
}
```

**Comment by andrewbranch** [comment]
Yeah, that’s the exact same issue.

> the type is wrong regardless

Not wrong, just not as specific as it could be. `42` is assignable to `{}`. `{}` just means not null or undefined. `42` is not null or undefined.

**Comment by ahejlsberg** [comment]
Actually, the issue reported [here](https://github.com/microsoft/TypeScript/issues/50706#issuecomment-1242366269) is a separate and unrelated issue. In 4.8, the check for `undefined` narrows the type of `T` to `T & ({} | null)`, but that should still be comparable to `42`.

**Comment by ahejlsberg** [comment]
I think the intent was to fix it in 4.8.4, but you're right, it isn't in that release. It's definitely fixed in 4.9.

@andrewbranch Looks like there was a failed cherry pick into 4.8.4, but not sure if that was the reason.
